src/System.Text.Primitives/System/Text/TextEncoding.cs

```diff
@@ -160,5 +220,203 @@ public enum EncodingName : byte
         public abstract bool TryEncode(string text, Span<byte> encodedBytes, out int bytesWritten);
 
         #endregion Encode Methods
+
+        #region Symbol Parsing / Formatting
+
+        public bool TryParseSymbol(ReadOnlySpan<byte> buffer, out Symbol symbol, out int bytesConsumed)
+        {
+            int trieIndex = 0;
+            int bufferIndex = 0;
+            bytesConsumed = 0;
+            while (true)
+            {
+                var node = _parsingTrie[trieIndex];
+                if (node.ValueOrNumChildren == 0)    // if numChildren == 0, we're on a leaf & we've found our value
+                {
+                    symbol = (Symbol)node.IndexOrSymbol;
+                    if (VerifySuffix(buffer, bufferIndex, symbol))
+                    {
+                        bytesConsumed = _symbols[node.IndexOrSymbol].Length - bufferIndex;
+                        return true;
+                    }
+                    else
+                    {
+                        symbol = 0;
+                        bytesConsumed = 0;
+                        return false;
+                    }
+                }
+                else
+                {
+                    int search = BinarySearch(trieIndex, bufferIndex, buffer[0]);    // we search the _parsingTrie for the nextByte
+
+                    if (search > 0)   // if we found a node
+                    {
+                        trieIndex = _parsingTrie[search].IndexOrSymbol;
+                        bytesConsumed++;
+                        bufferIndex++;
+                    }
+                    else
+                    {
+                        symbol = 0;
+                        bytesConsumed = 0;
+                        return false;
+                    }
+                }
+            }
+        }
+
+        public bool TryEncode(Symbol symbol, Span<byte> buffer, out int bytesWritten)
```

Comment:
The buffer is called "buffer" here and "encodedBytes" somewhere else. Let's make it consistent.